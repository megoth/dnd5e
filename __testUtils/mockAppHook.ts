import mockApp from "./mockApp";
import { AppModel } from "../src/models/app";

export default function mockAppHook(
  mock: jest.Mock,
  app = mockApp()
): AppModel {
  mock.mockReturnValue(app);
  return app;
}
