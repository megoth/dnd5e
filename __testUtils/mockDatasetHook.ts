import { mockSolidDatasetFrom, SolidDataset } from "@inrupt/solid-client";
import { SWRResponse } from "swr";

export default function mockDatasetHook(
  mock,
  presets: Partial<SWRResponse<SolidDataset, any>> = {}
) {
  return mock.mockReturnValue({
    data: mockSolidDatasetFrom("http://example.com"),
    error: null,
    ...presets,
  });
}
