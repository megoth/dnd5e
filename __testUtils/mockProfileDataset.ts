import {
  addStringNoLocale,
  addUrl,
  mockSolidDatasetFrom,
  mockThingFrom,
  setThing,
} from "@inrupt/solid-client";
import { vcard, foaf, rdf } from "rdf-namespaces";
import { chain } from "../src/utils";

export const aliceWebIdDatasetUrl = "http://example.com/alice";
export const aliceWebIdUrl = `${aliceWebIdDatasetUrl}#me`;
export const aliceName = "Alice";
export const aliceNick = "A";
export const alicePhoto = "http://example.com/alice.jpg";

export function mockProfileDataset(webId = aliceWebIdUrl) {
  const profile = chain(
    mockThingFrom(webId),
    (t) => addStringNoLocale(t, vcard.fn, aliceName),
    (t) => addStringNoLocale(t, vcard.nickname, aliceNick),
    (t) => addUrl(t, vcard.hasPhoto, alicePhoto),
    (t) => addUrl(t, rdf.type, foaf.Person),
    (t) => addUrl(t, foaf.openid, aliceWebIdUrl)
  );
  return chain(mockSolidDatasetFrom(aliceWebIdDatasetUrl), (d) =>
    setThing(d, profile)
  );
}
