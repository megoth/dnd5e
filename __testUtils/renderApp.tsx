import React from "react";
import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { LocalizationProvider, ReactLocalization } from "@fluent/react";
import { getFluentBundles } from "../src/models/language";
import { AppModel } from "../src/models/app";

export default function renderApp(
  app: AppModel,
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult {
  function rerender() {
    return render(
      <LocalizationProvider l10n={new ReactLocalization(getFluentBundles(app))}>
        {ui}
      </LocalizationProvider>,
      options
    );
  }
  const renderResult = rerender();
  return { ...renderResult, rerender };
}
