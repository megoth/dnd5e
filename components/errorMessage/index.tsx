import React from "react";
import { Localized } from "@fluent/react";
import NestedError from "nested-error-stacks";
import { getErrorId } from "../../src/models/error";

interface Props {
  error: NestedError;
}

export const TESTID_ERROR = "error";
export const TESTID_ERROR_TITLE = "error-title";

export default function ErrorMessage({ error }: Props) {
  const { message: errorUrl } = error;
  const errorId = getErrorId(errorUrl);
  return (
    <div data-testid={TESTID_ERROR}>
      <h1 data-testid={TESTID_ERROR_TITLE}>
        <span resource={errorUrl}>
          <Localized id={errorId}>{errorUrl}</Localized>
        </span>
      </h1>
    </div>
  );
}
