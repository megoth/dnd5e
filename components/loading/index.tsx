import React from "react";

export const TESTID_LOADING = "loading";

export default function Loading() {
  // Do not try to translate anything here - it's used when loading translations as well
  return <div data-testid={TESTID_LOADING}>Loading...</div>;
}
