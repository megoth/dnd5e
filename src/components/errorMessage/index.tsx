import React from "react";
import NestedError from "nested-error-stacks";

interface Props {
  error: NestedError;
}

export const TESTID_ERROR = "error";
export const TESTID_ERROR_TITLE = "error-title";

export default function ErrorMessage({ error }: Props) {
  // const { message: errorURL } = error;
  return (
    <div data-testid={TESTID_ERROR}>
      <h1 data-testid={TESTID_ERROR_TITLE}>
        {/*{translationURL ? (*/}
        {/*  <Localized id={getTranslationId(translationURL)}>*/}
        {/*    {errorURL}*/}
        {/*  </Localized>*/}
        {/*) : (*/}
        {error.message}
        {/*)}*/}
      </h1>
    </div>
  );
}
