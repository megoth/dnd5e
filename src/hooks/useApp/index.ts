import { useContext, useEffect } from "react";
import { AppContext } from "../../contexts/app";
import { appIsLoadingLocalizations } from "../../models/app";

export default function useApp(bundles: string[] = []) {
  const { app, setBundles } = useContext(AppContext);
  useEffect(() => {
    if (!app || !appIsLoadingLocalizations(app, bundles)) {
      return;
    }
    setBundles(bundles);
  }, [app, bundles, setBundles]);
  return app;
}
