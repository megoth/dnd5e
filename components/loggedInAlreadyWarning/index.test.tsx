import React from "react";
import * as solidUIReactFns from "@inrupt/solid-ui-react";
import LoggedInAlreadyWarning, {
  TESTID_LOGGED_IN_ALREADY_WARNING,
} from "./index";
import useDataset from "../../src/hooks/useDataset";
import mockDatasetHook from "../../__testUtils/mockDatasetHook";
import { mockProfileDataset } from "../../__testUtils/mockProfileDataset";
import {
  authenticatedWebId,
  mockUnauthenticatedSession,
} from "../../__testUtils/mockSession";
import useApp from "../../src/hooks/useApp";
import mockAppHook from "../../__testUtils/mockAppHook";
import { TESTID_ERROR } from "../errorMessage";
import { TESTID_LOADING } from "../loading";
import renderApp from "../../__testUtils/renderApp";

jest.mock("../../src/hooks/useApp");
const mockedAppHook = useApp as jest.Mock;

jest.mock("../../src/hooks/useDataset");
const mockedDatasetHook = useDataset as jest.Mock;

describe("LoggedInAlreadyWarning", () => {
  let app;

  beforeEach(() => {
    app = mockAppHook(mockedAppHook);
  });
  beforeEach(() =>
    mockDatasetHook(mockedDatasetHook, {
      data: mockProfileDataset(authenticatedWebId),
    })
  );

  it("renders", () => {
    const { asFragment, getByTestId } = renderApp(
      app,
      <LoggedInAlreadyWarning />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTID_LOGGED_IN_ALREADY_WARNING)).toBeDefined();
  });

  it("renders nothing if unauthenticated", () => {
    jest
      .spyOn(solidUIReactFns, "useSession")
      .mockReturnValue(mockUnauthenticatedSession());
    const { asFragment } = renderApp(app, <LoggedInAlreadyWarning />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders error if it fails to load profile", () => {
    mockDatasetHook(mockedDatasetHook, {
      data: null,
      error: new Error(),
    });
    const { getByTestId } = renderApp(app, <LoggedInAlreadyWarning />);
    expect(getByTestId(TESTID_ERROR)).toBeDefined();
  });

  it("renders error if it fails to load profile", () => {
    mockDatasetHook(mockedDatasetHook, {
      data: null,
      error: new Error(),
    });
    const { getByTestId } = renderApp(app, <LoggedInAlreadyWarning />);
    expect(getByTestId(TESTID_ERROR)).toBeDefined();
  });

  it("renders loading when fetching profile", () => {
    mockDatasetHook(mockedDatasetHook, {
      data: null,
    });
    const { getByTestId } = renderApp(app, <LoggedInAlreadyWarning />);
    expect(getByTestId(TESTID_LOADING)).toBeDefined();
  });
});
