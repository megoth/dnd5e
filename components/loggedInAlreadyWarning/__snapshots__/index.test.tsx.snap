// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoggedInAlreadyWarning renders 1`] = `
<DocumentFragment>
  <div
    data-testid="logged-in-already-warning"
  >
    <div
      class="bg-yellow-200 text-yellow-900 border-yellow-400 border px-2 py-1 rounded-sm mb-2"
    >
      LoggedInAlready
    </div>
  </div>
</DocumentFragment>
`;

exports[`LoggedInAlreadyWarning renders nothing if unauthenticated 1`] = `<DocumentFragment />`;
